### DISCUSSION TITLE
Raspberry Pi zero BLE issue

### DISCUSSION THREAD

**Discussion by patdemko** [mKeRix/room-assistant#578] [Q&A] [answered]
I'm trying to add two raspberry pi zero's to an existing cluster of two pi 3's and a pi 4 via ansible.  I started with a clean install of the latest raspberry pi OS lite image to an sd-card and then did an apt update & apt dist-upgrade.  Then I created / added ssh-keys to it and then pushed the configuration via ansible which I"m assuming installs all the latest dependencies and room assistant. Everything seemed to work (new pi's showed up in the cluster), but my BLE device never showed up as being closest to the new pi zero's.  I then stopped the service on one of the pi zero's and manually kicked it off with the verbose flags.  Here's what I'm seeing below in the output which seems to point to an issue with the BLE adapter since it says it's going into an unauthorized state. I next tried a rpi-update to that pi zero and tried it again but I'm still getting the same output in verbose mode.  I'm at a lost as to what to try next.  If anyone has any ideas, please let me know?

3/7/2021, 3:57:51 PM - debug - EntitiesService: Adding new entity ble-d8126338f2b.........406d-100-1
3/7/2021, 3:57:51 PM - debug - HomeAssistantService: Registering entity room-assistant-ble-d8126338f2b.........406d-100-1 under homeassistant/sensor/room-assistant/ble-d8126338f2b.........406d-100-1/config
3/7/2021, 3:58:20 PM - warn - BluetoothService: Did not detect any low energy advertisements in a while, resetting
3/7/2021, 3:58:20 PM - debug - BluetoothService: Resetting HCI adapter 0
3/7/2021, 3:58:25 PM - debug - BluetoothService: Adapter 0 went into state unauthorized
3/7/2021, 3:59:05 PM - warn - BluetoothService: Did not detect any low energy advertisements in a while, resetting
3/7/2021, 3:59:05 PM - debug - BluetoothService: Resetting HCI adapter 0
3/7/2021, 3:59:10 PM - debug - BluetoothService: Adapter 0 went into state unauthorized
3/7/2021, 3:59:50 PM - warn - BluetoothService: Did not detect any low energy advertisements in a while, resetting
3/7/2021, 3:59:50 PM - debug - BluetoothService: Resetting HCI adapter 0
3/7/2021, 3:59:55 PM - debug - BluetoothService: Adapter 0 went into state unauthorized
3/7/2021, 4:00:35 PM - warn - BluetoothService: Did not detect any low energy advertisements in a while, resetting
3/7/2021, 4:00:35 PM - debug - BluetoothService: Resetting HCI adapter 0
3/7/2021, 4:00:40 PM - debug - BluetoothService: Adapter 0 went into state unauthorized
3/7/2021, 4:01:20 PM - warn - BluetoothService: Did not detect any low energy advertisements in a while, resetting
3/7/2021, 4:01:20 PM - debug - BluetoothService: Resetting HCI adapter 0
3/7/2021, 4:01:25 PM - debug - BluetoothService: Adapter 0 went into state unauthorized

**Comment by mKeRix** [comment, marked as answer]
Did you run the `setcap` command from the [docs](https://www.room-assistant.io/integrations/bluetooth-low-energy.html#requirements)? The unauthorized state suggest that node does not have necessary permissions to start BLE scans. This may also happen if your NodeJS install was upgraded, since replacing a binary will cause its permissions to be lost. Try running the following command and then restarting:

```shell
sudo setcap cap_net_raw+eip $(eval readlink -f `which node`)
sudo setcap cap_net_admin+eip $(eval readlink -f `which hciconfig`)
```

**Reply by patdemko** [reply]
I guess I didn't and just assumed pushing it via ansible would run everything needed.  After running those commands I then started seeing my devices showing up on the raspberry pi zero's.  Although after I tried to push the latest config to make sure everything was in sync via ansible. I"m not getting errors off of those pi zero w's now on the downlaod NodeJS install script now.  I still need to dig into this and double check I didn't miss something else in the install doc but figured I'd let you know that the adapter part seems to work now.


TASK [nodejs : Download NodeJS install script] ******************************************************************************
skipping: [10.13.1.60]
skipping: [10.13.1.91]
skipping: [10.13.1.95]
skipping: [10.15.1.95]
fatal: [10.13.1.103]: FAILED! => {"changed": false, "dest": "/home/pi/install-node.sh", "elapsed": 0, "gid": 1000, "group": "pi", "mode": "0766", "msg": "Request failed", "owner": "pi", "response": "HTTP Error 404: Not Found", "size": 1944, "state": "file", "status_code": 404, "uid": 1000, "url": "https://raw.githubusercontent.com/sdesalas/node-pi-zero/master/install-node-v.lts.sh"}
fatal: [10.13.1.120]: FAILED! => {"changed": false, "dest": "/home/pi/install-node.sh", "elapsed": 0, "gid": 1000, "group": "pi", "mode": "0766", "msg": "Request failed", "owner": "pi", "response": "HTTP Error 404: Not Found", "size": 1944, "state": "file", "status_code": 404, "uid": 1000, "url": "https://raw.githubusercontent.com/sdesalas/node-pi-zero/master/install-node-v.lts.sh"}

**Reply by mKeRix** [reply]
It should, but the ansible playbook had a bug that would make this flaky. This has since been fixed - I'd recommend to `git pull` the ansible-playbooks repository and then re-run the playbook. That should also get ride of the errors you posted here, as it switches to a different installation process for NodeJS.

**Reply by patdemko** [reply]
So updating the ansible-playbooks and re-running it actually broke the Pi 3's & 4.  I ended up having to manually run the "sudo npm i --global --unsafe-perm room-assistant" command on each of those Pi's and then re-running the playbook got everything running again.  Thx for help!
